src/java.base/share/classes/com/sun/crypto/provider/HKDFKeyDerivation.java

```diff
@@ -0,0 +1,414 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import javax.crypto.KDFSpi;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.ByteArrayOutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import javax.crypto.KDFParameters;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * KDF implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC
+ * 5869.  This implementation provides the complete Extract-then-Expand HKDF
+ * function as well as Extract-only and Expand-only variants.
+ */
+abstract class HKDFKeyDerivation extends KDFSpi {
+
+    private final int hmacLen;
+    private final String hmacAlgName;
+
+    private static final int SHA256_HMAC_SIZE = 32;
+    private static final int SHA384_HMAC_SIZE = 48;
+    private static final int SHA512_HMAC_SIZE = 64;
+
+    // since we are using Arrays.binarySearch later, order is important
+    private static final int[] SUPPORTED_HMAC_SIZES = new int[]{
+            SHA256_HMAC_SIZE,
+            SHA384_HMAC_SIZE,
+            SHA512_HMAC_SIZE
+    };
+
+    /**
+     * The sole constructor.
+     *
+     * @param kdfParameters
+     *         the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *         if the initialization parameters are inappropriate for this
+     *         {@code KDFSpi}
+     */
+    private HKDFKeyDerivation(String hmacAlgName, int hmacLen,
+                              KDFParameters kdfParameters)
+            throws InvalidAlgorithmParameterException {
+        super(kdfParameters);
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(
+                    hmacAlgName + " does not support parameters");
+        }
+        // added to enforce valid values at reviewer's request
+        if (Arrays.binarySearch(SUPPORTED_HMAC_SIZES, hmacLen) < 0) {
+            throw new InternalError(
+                    "Subclass attempted to use an invalid hmacLen");
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
```

Comment:
Instead of doing a binary search whenever an HKDFKeyDerivation object is constructed, it is better to organize the algorithm and output length into an enum, this way, supporting new Hmac algorithms would require adding new enum value. This should be sufficient since these arguments are internally supplied. For example,
```
public enum SupportedHmac {
        SHA256("HmacSHA256", 32),
        SHA384("HmacSHA384", 48),
        SHA512("HmacSHA512", 64);

        public final String algo;
        public final int outLen;
        private SupportedHmac(String algo, int outLen) {
            this.algo = algo;
            this.outLen = outLen;
        }
 };
```

Reply:
Then the constructor can specify enum as the argument, e.g. 
```
    private HKDFKeyDerivation(SupportedHmac h,
                              KDFParameters kdfParameters)
            throws InvalidAlgorithmParameterException {
        super(kdfParameters);
        if (kdfParameters != null) {
            throw new InvalidAlgorithmParameterException(
                    h.algo + " does not support parameters");
        }
        this.hmacAlgName = h.algo;
        this.hmacLen = h.outLen;
    }
```